auth/src/krb.cpp

```diff
@@ -0,0 +1,942 @@
+#include <irods/rsAuthCheck.hpp>
+#include <irods/rsAuthResponse.hpp>
+#include <irods/rsAuthRequest.hpp>
+#endif // RODS_SERVER
+
+#include <gssapi.h>
+#include <openssl/md5.h>
+
+#include <string>
+
+static const int ikrbDebugFlag = 0;
+static const gss_OID gss_nt_service_name_krb = 0;
+static const unsigned int SCRATCH_BUFFER_SIZE = 20000;
+static char ikrbScratchBuffer[SCRATCH_BUFFER_SIZE];
+static int ikrbTokenHeaderMode = 1;  /* 1 is the normal mode,
+                                        0 means running in a non-token-header mode, ie Java; dynamically cleared. */
+
+#ifdef RODS_SERVER
+// =-=-=-=-=-=-=-
+// NOTE:: this needs to become a property
+// Set requireServerAuth to 1 to fail authentications from
+// un-authenticated Servers (for example, if the LocalZoneSID
+// is not set)
+//static const int requireServerAuth = 0;
+static int krbAuthReqStatus = 0;
+static int krbAuthReqError = 0;
+static const int krbAuthErrorSize = 1000;
+static char krbAuthReqErrorMsg[krbAuthErrorSize];
+#endif
+
+
+static rError_t *ikrb_rErrorPtr;
+
+// static gss_cred_id_t myCreds = GSS_C_NO_CREDENTIAL;
+static const int MAX_FDS = 32;
+static gss_ctx_id_t context[MAX_FDS] = {
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT,
+    GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT, GSS_C_NO_CONTEXT
+};
+
+static unsigned int context_flags;
+
+static gss_cred_id_t my_creds = GSS_C_NO_CREDENTIAL;
+
+namespace
+{
+    using json = nlohmann::json;
+    using log_auth = irods::experimental::log::authentication;
+
+    auto request(rcComm_t& comm, const json& msg)
+    {
+        auto str = msg.dump();
+
+        bytesBuf_t inp;
+        inp.buf = static_cast<void*>(const_cast<char*>(str.c_str()));
+        inp.len = str.size();
+
+        bytesBuf_t* resp{};
+        auto ec = procApiRequest(&comm, 110000, static_cast<void*>(&inp), nullptr, reinterpret_cast<void**>(&resp), nullptr);
+
+        if (ec < 0) {
+            THROW(ec, "failed to perform request");
+        }
+
+        return json::parse(static_cast<char*>(resp->buf), static_cast<char*>(resp->buf) + resp->len);
+    } // request
+
+    // TODO: std::vector could be a const ref
+    auto verify(const json& p, std::vector<std::string> n)
+    {
+        for(auto&& x : n) {
+            if(!p.contains(x)) {
+                return std::make_tuple(false, x);
+            }
+        }
+
+        return std::make_tuple(true, std::string{});
+    } // verify
+
+    void krb_log_error_1(rError_t* _r_error,
```

Comment:
Do we expect `_r_error` to ever be null?

If not, we should pass by reference. If we're unsure, leave it as is. This type of thing is not a blocker.

Reply:
I will look into making this a reference.

Reply:
Done